### ISSUE TITLE
Cannot launch demo

### ISSUE THREAD

**Issue by yoann-solana** [SCU-RSL-ROS/radar_omnipresense#11]
Hi,
I am trying to test the OPS241-A Short Range Radar Sensor on Ubuntu 16.04 with ROS Kinetic but I cannot launch the single or multi radar launch demo.

I followed the procedure in the ReadMe (rapidJson installed and ROS (working with other projects)).

This device is plugged and available at /dev/ttyACM0, I can read the data from a serial console.

Attached, this is the error I got, when I launch the single_radar.launch file.
[error_single_radar.txt](https://github.com/SCU-RSL-ROS/radar_omnipresense/files/2323614/error_single_radar.txt)

Did I forgot something in the setup? Can you fix this please?

Regards,
Yoann.

**Comment by jimwhitfield** [comment]
Hi Yoann, The rapidjson library became quite a thorn in this project because the different ubuntus had vastly different rapidjson library versions.  (This is why we ultimately included one frozen in time.)  It's possible you were affected, too.   I can't say for certain until I see what failed to parse.  Hopefully the log file mentioned in the log file you included will give more insight.  
Your included log mentioned: 
 Assertion `IsString()' failed.
....radar_omnipresense/radar_publisher __name:=radar_data_and_services __log:=**/home/yoyo/.ros/log/daeb17c4-a9e2-11e8-9c07-3052cb9a765b/radar_data_and_services-1.log**
It would be most helpful to see the related (probably the ending) contents of that .log file to see what was not a string.  (I presuming running again will end up with a different log file path if that one is not available.)

**Comment by yoann-solana** [comment]
Hi,
Thank you for the help. 
I have attached the log file. However, inside this file nothing refer to the string which is parsed.
[radar_data_and_services-2-stdout.log](https://github.com/SCU-RSL-ROS/radar_omnipresense/files/2328896/radar_data_and_services-2-stdout.log)

Regards,
Yoann

**Comment by jimwhitfield** [comment]
Hm, you're right. OK, so, most of the last checkins were all pertaining to getting the ROS build farm to compile the driver (and not always successfully)  
I think the most recent checkin I made, 7b6e889f7c4cf48eb3902e20ed6017af6984fbd0 might be before we were trying so hard for that.  
Details: Ubuntu 16.04 had an ancient release called rapidjson-dev  (pre 1.0) Garren had developed against a more recent release and then had to back out features when we wanted to build/run against a clean 16.04.    Then we found out 18.04 had a much more advanced version of rapidjson that broke backward compatibility.   If my memory serve correctly (and probably doesn't)  your error makes me recall the fun we had with 18.04.  Are you using Ubuntu 18.04?  
(Ultimately, we put in our own copy in the repo.  I wonder if paths are such that you're using the system rapidjson and not the one from the repo?)   
I'll try to build and run what you're running once I know more about your platform   Which Ubuntu?  Are you using a pi or linux box (or windows? or mac?)  Any other tips for making a platform like yours?
Are you comfortable making some changes to C code to perhaps do more logging?

**Comment by yoann-solana** [comment]
Hi,

> Any other tips for making a platform like yours?

I am using Ubuntu 16.04 64 bits with ROS kinetic. In my workspace I have a clone of this repository and also a clone of rapidJson v1.1.0. 

> Are you comfortable making some changes to C code to perhaps do more logging?

Yes, I can make some change in the code and do more logging. 

Below, this is the procedure I followed to install and play with the radar.
For rapidJson:
```
cd path_rapidJson
mkdir build && cd build
cmake ..
make
sudo make install
```
For radar_omnipresense
```
cd root_of_catkin_workspace
catkin build
```
To launch the demo
```
source catkin_ws/devel/setup.bash
roslaunch radar_omnipresense single_radar.launch
```
Did you see something wrong with this procedure?

I have done the previous procedure with the version [v1.0.0](https://github.com/Tencent/rapidjson/tree/v1.0.0) of rapidJson (22 april 2015, sha = 73063f5002612c6bf64fe24f851cd5cc0d83eef9), but I have the same error and a similar log file. In my system the rapidjson-dev package is not installed only the one from the catkin_ws.

Do you have an install procedure using the rapidJson in your repository?

Thank you again for the help.
Regards,
Yoann.

**Comment by jimwhitfield** [comment]
The procedure you described looks right.  
(I don't have a pi or robot with me this morning, so I'll have to try to reproduce this later today.  Is your hardware a standard PC or a raspberry pi?)
  
The line in your initial log 
`/home/yoyo/ROS/ws_sensors/src/radar_omnipresense/src/../lib/rapidjson/document.h:1771`
makes me realize you _are_ using the library delivered as part of the repo.  Hmm, the mystery deepens.

Code-wise: it would be helpful to see what is being passed into the json parser to see what is giving it trouble.  You might want to log what comes in from the sensor before it goes to the json library.  (Or if possible, catch the exception.)    
Sorry this is being such a pain.  rapidjson seemed like a very good choice but it has caused more hassle than ever anticipated.

**Comment by yoann-solana** [comment]
Hi,
Sorry I forgot to write it. I am working on a PC not a Raspberry Pi.

No problem, for this issue, I understand that sometime we can have troubles with external libraries.

I did not have time to check at the source code, but do you have an idea where I can look (file, function?) in order catch the string before to be passed to the rapidJson library?

Thanks in advance,
Yoann.

**Comment by yoann-solana** [comment]
Hi,
Ok, I did not have time to check the code yet. But if it is necessary, I will think to this alternative to use the sensor on ROS.
Thank you for the time,
Regards,
Yoann.

**Comment by jimwhitfield** [comment]
Everything is checked in.
As soon as I get release rights, this will be on the build farm.
Until then, you have to download the code from this repo and build it.

**Comment by yoann-solana** [comment]
Hi,
Sorry for the delay. 
The version of the sensor is OPS241-A. Is it compatible with the last version?
Regards,
Yoann.

**Comment by yoann-solana** [comment]
Hi,
I just test the OPS241-A sensor with the last commit ae936332a95f4be6a5b548741271cc3537bb2d41 in master and it works. Thank you for the help.

I have more questions about the data, I am not sure if I can ask in this issue.

- I noticed a weird thing, without nothing in the field of view (FoV) of the sensor except the ceiling. The reported speed is always around 0.64. Do I need to configure something or make a calibration? Is the reported speed is in m/s?
- Could it be possible to have the angular position and range of the target? Not only the magnitude and direction of the speed?
- Last question, what is the meaning of objnum? Is it always 1? I though it was the number of target in the FoV but a test with several target moving in the FoV did not change this value.

Below is a copy of a rostopic echo command. As you can see, with nothing moving in the FoV the speed is not null.
```
---
speed: 0.639999985695
direction: inbound
time: 1882.98596191
sensorid: /dev/ttyACM0
objnum: 1.0
metadata: 
  seq: 0
  stamp: 
    secs: 1538119067
    nsecs: 476822811
  frame_id: ''
fft_data: 
  real: []
  imaginary: []
raw_data: 
  i: []
  q: []
---
```

Thanks again for the help.
Yoann

**Comment by jimwhitfield** [comment]
Hi Yoann
Thank you for getting back to us on this issue.
1) units:  unless you sent some other request (like "US" for mph or "UK" for km/h) the numbers are in meters per second.   The API is availavble at https://www.omnipresense.com/wp-content/uploads/2018/09/AN-010-E_API_Interface.pdf
1a) spurious data.  perhaps there is a fan in the ceiling?  Another option is to increase the squelch in case it's a radio-noisy environment.  Q_ is the prefix.  QI is the default QX sets a threshold 10 times higher.

2) Sorry, it's a one TX one RX radar with doppler processing.  Physics limits it to speed and in/outbound direction.

3) Objnum is currently always one.  Version 1.2 of the firmware brings sensor-calculated "top-N readings" (as opposed to "the top speed" which is the default and only mode available in firmware 1.1)  Correlation algorithms can associate the readings from one beat to the next, but at the moment, we only have that in python code, not C code.  Please reach out to OmniPreSense via customerservice@omnipresense.com if you think you might want to make use of this.

We welcome public questions were in the repo and directly.  That said, I think this particular ticket is resolved and I'll close the issue.  Feel free to open new issues as needed.  Thank you!
